pkg/distributor/influx.go

```diff
@@ -0,0 +1,177 @@
+		return err
+	}
+
+	// Sigh, a write API optimisation needs me to jump through hoops.
+	pts := make([]mimirpb.PreallocTimeseries, 0, len(ts))
+	for i := range ts {
+		pts = append(pts, mimirpb.PreallocTimeseries{
+			TimeSeries: &ts[i],
+		})
+	}
+
+	level.Debug(spanLogger).Log(
+		"msg", "Influx to Prometheus conversion complete",
+		"metric_count", len(ts),
+	)
+
+	req.Timeseries = pts
+	return nil
+}
+
+// InfluxHandler is a http.Handler which accepts Influx Line protocol and converts it to WriteRequests.
+func InfluxHandler(
+	maxRecvMsgSize int,
+	requestBufferPool util.Pool,
+	sourceIPs *middleware.SourceIPExtractor,
+	retryCfg RetryConfig,
+	push PushFunc,
+	_ *PushMetrics, // TODO(alexg) add pushMetrics()
+	_ prometheus.Registerer, // TODO(alexg): add reg
+	logger log.Logger,
+) http.Handler {
+	//TODO(alexg): mirror otel.go implementation where we do decoding here rather than in parser() func?
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := utillog.WithContext(ctx, logger)
+		if sourceIPs != nil {
+			source := sourceIPs.Get(r)
+			if source != "" {
+				logger = utillog.WithSourceIPs(source, logger)
+			}
+		}
+
+		supplier := func() (*mimirpb.WriteRequest, func(), error) {
+			rb := util.NewRequestBuffers(requestBufferPool)
+			var req mimirpb.PreallocWriteRequest
+
+			if err := parser(ctx, r, maxRecvMsgSize, rb, &req, logger); err != nil {
+				// TODO(alexg): Do we even need httpgrpc here?
+				// Check for httpgrpc error, default to client error if parsing failed
+				if _, ok := httpgrpc.HTTPResponseFromError(err); !ok {
+					err = httpgrpc.Error(http.StatusBadRequest, err.Error())
+				}
+
+				rb.CleanUp()
+				return nil, nil, err
+			}
+
+			cleanup := func() {
+				mimirpb.ReuseSlice(req.Timeseries)
+				rb.CleanUp()
+			}
+			return &req.WriteRequest, cleanup, nil
+		}
+		req := newRequest(supplier)
+		// https://docs.influxdata.com/influxdb/cloud/api/v2/#tag/Response-codes
+		if err := push(ctx, req); err != nil {
+			if errors.Is(err, context.Canceled) {
+				level.Warn(logger).Log("msg", "push request canceled", "err", err)
+				w.WriteHeader(statusClientClosedRequest)
+				return
+			}
+			if errors.Is(err, io2.ErrReadLimitExceeded) {
+				// TODO(alexg): One thing we have seen in the past is that telegraf clients send a batch of data
+				// if it is too big they should respond to the 413 below, but if a client doesn't understand this
+				// it just sends the next batch that is even bigger. In the past this has had to be dealt with by
+				// adding rate limits to drop the payloads.
+				level.Warn(logger).Log("msg", "request too large", "err", err)
+				// TODO(alexg): max size and bytes received in error?
+				w.WriteHeader(http.StatusRequestEntityTooLarge)
+				return
+			}
+			// From: https://github.com/grafana/influx2cortex/blob/main/pkg/influx/errors.go
+
+			var (
+				httpCode int
+				errorMsg string
+			)
```

Comment:
While the code below isn't extremely complex, it's still a couple of `if`s, so I'd write here some sane defaults.
```suggestion
			httpCode := http.StatusInternalServerError
			errorMsg := "unknown error"
```

Reply:
Agreed. Done!